### ISSUE TITLE
Identify & Classify Singleton Use in Gerbera

### ISSUE THREAD

**Issue by elmodaddyb** [gerbera/gerbera#374]
The gerbera system has several singleton classes which provide isolation of capability.  However, it produces artificial relationships between objects that can be more concrete by linking objects through memory references.

* The issue identifies all use of `Singleton`
* The issue finds all related classes to understand the sharing of the `Singleton`
* The issue identifies candidate singletons that can be implemented without the singleton pattern.

**Comment by elmodaddyb** [comment]
`ConfigManager` singleton is used by `upnp_xml` but it is only called from `UpnpXML_RenderDeviceDescription` which is called from `server`.

> We can instead pass the `ConfigManager` from main server and make `upnp_xml` more stateless

**Comment by v00d00** [comment]
Yeah this was the motivation towards starting the UpnpXMLBuilder changes.

We can pass some information into the UpnpXMLBuilder constructor now.

Prime ones being the Device description string and the server bound hostname string.

That decouples those objects from knowing that config even exists which is nice.

**Comment by pamapa** [comment]
Singletons are not that bad. I have seen a lot of code which uses the singleton pattern.

The whole idea with the singleton manager is a graceful shutdown, which is required e.g. to avoid corrupt DBs.

When i look at the code we could improve the usage of it though.

Idea:
Instead of accessing allover with X.getInstance(), we could simply inject the object via constructor and then use it like _cm.getIntOption(). Like this the "using classes" do not need to know, that they deal with a singleton at all. This idea is inspired by dependency injection.

Pseudo code:
```
class ISingleton {
...
}

class ConfigManager : public ISingleton {
...
}

class ClassWhichNeedsConfigManager {
private:
    std::shared_ptr<ConfigManager> _cm;

public:
    ClassWhichNeedsConfigManager(std::shared_ptr<ConfigManager> cm) : _cm(cm) {
    }
}
```

For cases where a singleton needs another singleton, the singleton manager shell handle this and inject the dependency into the singleton.

The names of the singletons are not optional. e.g. Timer is in fact a manager of notify events.
PlayHook can be eliminated by adding the trigger function directly to the ConfigManager.

**Comment by whyman** [comment]
I think the singletons are still bad. Especially when they are globally accessed, no sensible way to control the lifetimes of the objects.

Having a single instance and passing it into the constructor is the way to go. Nothing fancy required 👍

**Comment by pamapa** [comment]
i guess this issue can be closed (fixed via #573)

**Comment by whyman** [comment]
Yes true :tada:
